### ISSUE TITLE
No package in Fedora repositories

### ISSUE THREAD

**Issue by souryogurt** [cgreen-devs/cgreen#225]
It would be great if such an awesome framework will be available from Fedora's repositories! :)

## Steps to reproduce
1. Install Cgreen for development `sudo dnf install cgreen-devel cgreen-runner`

## Actual Result
```console
No match for argument: cgreen-devel
Error: Unable to find a match: cgreen-devel
```

## Expected result
The `dnf` installs Cgreen library itself, development files and `cgreen-runner`.

**Comment by souryogurt** [comment]
Hello there! :) My name is Egor, and I just C enthusiast and using Fedora and Cgreen at home for pet projects. 
I am going to pack and maintain the Cgreen package for Fedora Linux. 
It seems to build and pack Cgreen 1.2.0 on all primary Fedora's architectures these issues must be resolved:
- #218 
- #212 
- #221 
- #211 
- #223

For the first RPM release, I'll use [cgreen v1.2.0](https://github.com/cgreen-devs/cgreen/releases/tag/1.2.0) as available on the release page of this repository with patches applied for listed issues.

**Comment by thoni56** [comment]
Hi Egor! And thank you for volouteering! That is much appreciated. 

I suggest that we make a new 1.3.0 release that contains the mentioned, and any other fixes that you might find, too make that work easier.

Please go ahead with any further tests so we can discover any outstanding issues. Then I’ll make a new release.

**Comment by souryogurt** [comment]
I guess making a new 1.3.0 release before passing the review process for a package according to maintainer's guidelines will not be anyhow easier anyway because it is the first time submission. Almost all patches for issues mentioned above are already accepted by you, so patches shipped with RPM are the same as in the master branch.
There is only one patch left that not merged - #224. It is a patch that fixes a small issue in the test that assumes that the test is always compiled with char defined as `signed`, that not the case for ARM architectures.

Since there is a review process in Fedora's community, the listed patches with links to issues in upstream's tracker might even help to pass it, since they demonstrate for reviewers that I working with you(upstream) to resolve them and that I have strong intentions to continue to support package in the future. :)

So probably, for easier the submission of the first version of RPM package, it would be simpler to submit a current latest release with patches, pass the review and then just update the approved package using a new version of source tarball when you deem necessary to release it. So actually the are no blockers to submit RPM to fedora right now. 

But if you want, I can wait for v1.3.0 release and submit the package only after that.

**Comment by thoni56** [comment]
Good point, showing that there are people working on it. Then please, go ahead.

For issue #224 the patch obviously fixes the problem, I'm just pondering if it is the correct solution. It might point to a deeper issue about "bytes"... I'll get back to you on that shortly.

**Comment by thoni56** [comment]
Should we create a repo for Fedora-packaging, as we did for Debian in #208 ? Is there a difference except for procedures?

**Comment by souryogurt** [comment]
@thoni56 RPM's `spec` files are hosted and maintained on Fedora's https://src.fedoraproject.org/  server after approvement. So it is not required to maintain this repo in `cgreen-devs's` repo. 
But I can upload them to `cgreen-devs's` repo if you need these files. 

Actually I already uploaded them temporarily[ to this repo](https://github.com/souryogurt/cgreen-rpm) to give access for spec file for reviewers during spec review.

**Comment by thoni56** [comment]
Right. Good. As long as there is stable repo where the files are stored so that a future maintainer can easily pick them up and continue maintenance. And I suppose we should make a link in the Readme about that when they are approved. (or maybe we should "announce" the temporary repo for now, it might take a while to approve and somebody might want to try out the RPM-packaging?)

**Comment by thoni56** [comment]
Just noticed that your initial comment mentioned `cgreen-runner` and `cgreen-devel`. Would `cgreen-runner` be the runner and the library? and `cgreen-devel` the required source files?

**Comment by souryogurt** [comment]
@thoni56 There are should be three packages according to [Fedora's guidelines](https://docs.fedoraproject.org/en-US/packaging-guidelines/#_devel_packages):
- `cgreen` that contains versioned `libcgreen.so.1.2.0` shared library;
- `cgreen-devel` that contains development files: headers,  compiled  HTML page with a manual, and symbolic link to an actual library that is used at link time: `libcgreen.so` and `libcgreen.so.1`. In other words: the files required for developing application using Cgreen;
- `cgreen-runner` package that contains `cgreen-runner` and `cgreen-debug`  executables. as well as corresponding manual pages.

The `cgreen-devel` and `cgreen-runner` packages will depend on `cgreen` package as required by [fedora's guidelines](https://docs.fedoraproject.org/en-US/packaging-guidelines/#_devel_packages). So, the `cgreen` package with the actual library will be installed automatically, when the user will install `cgreen-devel` or `cgreen-runner`.

So the end-user will install `cgreen-devel` on his work machine to develop apps with test that use Cgreen. He also can install `cgreen-runner` if he wants to run tests using `cgreen-runner` app. 

I'm not sure that this should be noticed somehow. Probably it is better to notice it when the package will be approved and will actually come to the fedora repository. But maybe it worse to add the link to this issue in readme just to allow other developers to track the status or contribute.
But if someone are interested you might install binaries from COPR right now: 
just:
```
dnf copr enable souryogurt/cgreen
dnf install cgreen-devel cgreen-runner`
```

Probably there are already maintainers that packaging `cgreen` for Fedora, then please check what already done here: https://github.com/souryogurt/cgreen-rpm

I'm a bit late with posting a package to review because we have come back to work at the office here at Russia, but I'm working on it. I'll post the link to the review request here when it is ready. But if you are (maintainers), are faster then me, then please go on :)

**Comment by souryogurt** [comment]
Added review request: https://bugzilla.redhat.com/show_bug.cgi?id=1843300

**Comment by thoni56** [comment]
Thanks for the explanation, that sounds good. Added a link to this issue in the README.

Also thanks for the review request. Looking forward to seeing it pass ,-)

**Comment by thoni56** [comment]
BTW, I felt the need to create a 1.3.0 release. In one of my projects I discovered that the CgreenValueType enums clashed with potentially commonly used values, in particular `DOUBLE`.

I hope that does not interfere with this process. If so, we could retract the release, but I suppose the whole idea for distributions packaging strategies is to isolate development of the source/upstream software with the packages in the distribution.

**Comment by souryogurt** [comment]
I have upgraded the package to 1.3.0 in august. The package is approved already, but I am still looking for a sponsor. Fedora community has a sponsorship model for new packagers, so to submit an approved package to the Fedora repository, someone must add me to the packagers group.

Trying to contribute to other Fedora packages to get the sponsorship, and also just asked (15 November) to sponsor me. Hope It will happen soon :)

**Comment by souryogurt** [comment]
Finally, I was approved to Fedora's packager's group, so no can submit packages to the central Fedora repository!:) You can not even imagine how happy I am about this:)
I submitted a request to create Cgreen repo for the package. It will be reviewed in 24 hours by admins, and I hope such an excellent framework as Cgreen will be available from the primary Fedora's repos soon :)))

https://pagure.io/releng/fedora-scm-requests/issue/31333

**Comment by souryogurt** [comment]
Ok, the repo for spec is created in fedora's repo: https://src.fedoraproject.org/rpms/cgreen

I already build it for Fedora Rawhide, 33 and 32 releases.

Into Fedora Rawhide and all next Fedora releases, it will come tomorrow! :)

However, to submit the Cgreen and make it available for the current active Fedora 33 and Fedora 32 e users, the package should pass an additional QA step.

I have created two requests for QA. Below are the URLs to track a status:
https://bodhi.fedoraproject.org/updates/FEDORA-2020-93abc90e56
https://bodhi.fedoraproject.org/updates/FEDORA-2020-7f9fd147ff

In Fedora's update policy, the package should get one :+1: like from testers or just wait 7 days before it will be automatically marked as stable by the system. When it becomes marked as stable, the system will push it to current Fedora users next Tuesday.

So I guess we need to wait around 14 days, and anyone will be able to install Cgreen not only in Rawhide but also in Fedora 33 and Fedora 32! :)

**Comment by thoni56** [comment]
Good news! Thank you for taking a lead on this!

**Comment by souryogurt** [comment]
The Cgreen has reached Fedora 32, 33, and Rawhide repositories! So I guess this issue can be closed.

**Comment by thoni56** [comment]
Yes it can! Thank you, Egor (@souryogurt), for your efforts! And Merry Christmas!
